Update Rust documentation with workflow speed improvements

tgross35 opened the pull request:
This updates instructions for downloading the Rust source tree to use the tarball rather than cloning the repository.

I'll submit this to the list but a quick review here is nice :)

tgross35 commented:
Talked [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Shallow.20submodules) and the team recommended using the source tarball rather than cloning via git. This is significantly faster compared to the time to clone, and it only keeps the `library/` directory so space usage is much better.

Technically `rustup componnet add rust-src` adds the entire source tree. I can adjust it to do this if needed, but I don't think anything outside of library is used.

I have tested this by using the [standalone installer](https://forge.rust-lang.org/infra/other-installation-methods.html#standalone-installers) in a docker container then runninng the command, default build with Rust succeeds.

ojeda commented:
> Update sample 'make' commands to use '-j$(nproc)' to run parallel builds

Originally, I intentionally did not write extra options, because the idea was to show what was required, not anything extra.

In other words, the document is not intended to be a general guide to kernel development, or to replace the kernel documentation elsewhere nor the GNU Make documentation; in order to keep it minimal and focused on the Rust bits.

tgross35 commented:
Understood, I can remove that.

Are you ok with switching to the tarball?

ojeda commented:
Unsure -- can you think of some reason not to do it? For instance, is it easier/harder when updating? I guess in both cases the user can simply delete the folder and reapply the instructions, right?

By the way, the sentence "...upgrading the Rust compiler version later on requires manually updating this clone." would need to be slightly tweaked too if we do it.

tgross35 commented:
> Unsure -- can you think of some reason not to do it?

Nope, but I thought maybe you could :) It seems straightforward enough and I think that is how `rustup` does it

> By the way, the sentence "...upgrading the Rust compiler version later on requires manually updating this clone." would need to be slightly tweaked too if we do it.

Thanks, I will update this when I fix the other thing

bjorn3 commented:
There are two separate tarballs. One with the source code of the entirety of rustc which you can use to build rustc from scratch ([rustc-1.71.0-src](https://static.rust-lang.org/dist/rustc-1.71.0-src.tar.gz)) and one which contains just enough to build the standard library ([rust-1.71.0-src](https://static.rust-lang.org/dist/rust-src-1.71.0.tar.gz)). It is the later which ends up providing the files in `lib/rustlib/src/rust`. It has a different layout from rustc-src though.

tgross35 commented:
> There are two separate tarballs. One with the source code of the entirety of rustc which you can use to build rustc from scratch ([rustc-1.71.0-src](https://static.rust-lang.org/dist/rustc-1.71.0-src.tar.gz)) and one which contains just enough to build the standard library ([rust-1.71.0-src](https://static.rust-lang.org/dist/rust-src-1.71.0.tar.gz)). It is the later which ends up providing the files in `lib/rustlib/src/rust`. It has a different layout from rustc-src though.

That one is much smaller, I'll update this to use that one. Thanks for the link

tgross35 commented:
I removed mentions of `-j$(nproc)` and updated to use the smaller source tarball. Great change, it takes about 5 seconds to download instead of about 8 minutes for the original `git clone`.

I can submit this to the list if there aren't any further changes needed.

ojeda commented:
Sounds good. (To be clear, reviews typically happen in the list, i.e. you may get more feedback there.)

I would recommend splitting the changes into two different patches, since they are independent. Also, I would explain the improvement in time in the commit message.

With an standalone installer, is the sysroot the same when you install a different version? I am asking because if the sysroot stays the same in some cases (unlike e.g. `rustup` ones), the clone would have the advantage of making it a bit easier to switch versions (i.e. just checking out the other tag).

ojeda commented on Documentation/rust/quick-start.rst:
```diff
@@ -56,16 +58,16 @@ If ``rustup`` is being used, run::
-		--branch $(scripts/min-tool-version.sh rustc) \
-		https://github.com/rust-lang/rust \
-		$(rustc --print sysroot)/lib/rustlib/src/rust
+	curl -L https://static.rust-lang.org/dist/rust-src-$(scripts/min-tool-version.sh rustc).tar.gz | \
```
Is `-L` needed? Should we have other options, like `--proto '=https' --tlsv1.2 -sSf`?

Also, is the `\` at the end needed?

bjorn3 replied:
Without `-L` it would break if this url changes into a redirect in the future. The `\` at the end is necessary to prevent a shell from trying to run the partial command at the newline rather than interpreting everything as a single command.

ojeda replied:
> Without `-L` it would break if this url changes into a redirect in the future

Yeah, I meant whether there was a redirect going on now already.

I guess the `-L` won't hurt, but hopefully the URL won't change.

> The `\` at the end is necessary to prevent a shell from trying to run the partial command at the newline rather than interpreting everything as a single command.

The `|` already does that, no? (even in the interactive case)

bjorn3 replied:
I expected `|` at the end of a line to be an error in non-interactive cases, but it seems that it works just fine.

ojeda replied:
Yeah, in my case I used to put the `|` in the next line, so that it looked like what one does in some functional languages. But since the `\` could be avoided, I started putting the `|` at the end in the previous line instead.

tgross35 replied:
I have updated this to drop the `\` but kept the `-L`, just in case

tgross35 commented:
> With an standalone installer, is the sysroot the same when you install a different version? I am asking because if the sysroot stays the same in some cases (unlike e.g. rustup ones), the clone would have the advantage of making it a bit easier to switch versions (i.e. just checking out the other tag).

It seems like they do not change by version, testing on debian the system package manager uses `/usr/lib/rustlib/src` and the standalone installer uses `/usr/local/lib/rustlib/src`. I updated the documentation to just suggest removing `/lib/rustlib/src/rust` within the rust sysroot, which is sufficient (verified the command works again after removing just that directory).

I will post this to the list shortly.

edit: all done https://lore.kernel.org/rust-for-linux/20230718054416.861412-1-tmgross@umich.edu/T/#t

ojeda commented:
Thanks! I see it in Lore now (it took a while to arrive, it seems).

Looks quite good for a first submission!
